tests/components/samsungtv/test_media_player.py

```diff
@@ -55,250 +119,417 @@ class UnhandledResponse(Exception):
+async def test_update_off(hass, remote):
+    """Testing update tv off."""
+    await setup_samsungtv(hass, MOCK_CONFIG)
+    state = hass.states.get(ENTITY_ID)
+    remote.control = mock.Mock(side_effect=OSError("Boom"))
+    # await async_update()
+    assert STATE_OFF == state.state
+
+
+async def test_send_key(hass, remote, wakeonlan):
+    """Test for send key."""
+    await setup_samsungtv(hass, MOCK_CONFIG)
+    state = hass.states.get(ENTITY_ID)
+    assert await hass.services.async_call(
+        DOMAIN, SERVICE_VOLUME_UP, {ATTR_ENTITY_ID: ENTITY_ID}, True
+    )
+    # key and async_update called
+    assert remote.control.call_count == 2
+    assert remote.control.call_args_list == [call("KEY_VOLUP"), call("KEY")]
+    assert STATE_ON == state.state
+
+
+async def test_send_key_autodetect_websocket(hass):
+    """Test for send key with autodetection of protocol."""
+    with patch("samsungctl.Remote") as remote, patch(
+        "homeassistant.components.samsungtv.media_player.socket"
+    ):
+        await setup_samsungtv(hass, MOCK_CONFIG_AUTO)
+        state = hass.states.get(ENTITY_ID_AUTO)
+        assert await hass.services.async_call(
+            DOMAIN, SERVICE_VOLUME_UP, {ATTR_ENTITY_ID: ENTITY_ID_AUTO}, True
+        )
+        assert remote.control.call_count == 1
+        assert remote.call_args_list == [call(AUTODETECT_WEBSOCKET)]
+        assert STATE_ON == state.state
+
+
+async def test_send_key_autodetect_legacy(hass):
+    """Test for send key with autodetection of protocol."""
+    with patch(
+        "samsungctl.Remote", side_effect=[OSError("Boom"), mock.DEFAULT]
+    ) as remote, patch("homeassistant.components.samsungtv.media_player.socket"):
+        await setup_samsungtv(hass, MOCK_CONFIG_AUTO)
+        state = hass.states.get(ENTITY_ID_AUTO)
+        assert await hass.services.async_call(
+            DOMAIN, SERVICE_VOLUME_UP, {ATTR_ENTITY_ID: ENTITY_ID_AUTO}, True
         )
-        self.device.get_remote = mock.Mock(return_value=_remote)
-        command = "HELLO"
-        self.device.send_key(command)
-        self.assertEqual(STATE_ON, self.device._state)
-        # verify that _remote.control() get called twice because of retry logic
-        expected = [mock.call(command), mock.call(command)]
-        assert expected == _remote.control.call_args_list
-
-    def test_send_key_unhandled_response(self):
-        """Testing unhandled response exception."""
-        _remote = mock.Mock()
-        _remote.control = mock.Mock(
-            side_effect=self.device._exceptions_class.UnhandledResponse("Boom")
+        assert remote.call_count == 2
+        assert remote.call_args_list == [
+            call(AUTODETECT_WEBSOCKET),
+            call(AUTODETECT_LEGACY),
+        ]
+        assert STATE_ON == state.state
+
+
+async def test_send_key_autodetect_none(hass):
+    """Test for send key with autodetection of protocol."""
+    with patch("samsungctl.Remote", side_effect=OSError("Boom")) as remote, patch(
```

Comment:
Why can't we use the `remote` fixture here?

Reply:
Wasn't able to overwrite the `side_effect` from the fixture.

Reply:
Yeah, if we need to control a side effect or return value of the `Remote` class mock, we can't use the fixture as is. I see three options:

1. Patch in the test and don't use the fixture.
2. Make a new fixture that sets a side_effect, and use this fixture instead.
3. Change the existing fixture to instead of yielding the remote instance yield the mocked remote class. Then we can set a side_effect in the test as needed. We then need to change the tests to check the return_value of the fixture value instead of the fixture value, since the fixture value will be the remote class and not the remote instance.